Detectors/EMCAL/simulation/include/EMCALSimulation/Patches.h

```diff
@@ -0,0 +1,129 @@
+struct FastOrStruct {
+  std::vector<double> mADCvalues;
+  double mPreviousTimebinADCvalue = 0.;
+
+  /// Compute current timesum
+  double timesum()
+  {
+    double timesumvalue = 0;
+    for (auto ADCvalue : mADCvalues) {
+      timesumvalue += ADCvalue;
+    }
+    return timesumvalue;
+  }
+
+  void init()
+  {
+  }
+
+  // Update internal ADC values (4 timebins)
+  void updateADC(double ADCvalue)
+  {
+
+    if (mADCvalues.size() == 4) {
+      mPreviousTimebinADCvalue = mADCvalues.front();
+      mADCvalues.erase(mADCvalues.begin());
+    }
+    mADCvalues.push_back(ADCvalue);
+  }
+};
+
+/// \struct Patches
+/// \brief Patches creator, based on the Patches
+/// \ingroup EMCALsimulation
+/// \author Markus Fasel, ORNL
+/// \author Simone Ragoni, Creighton U.
+/// \date 03/12/2022
+///
+
+struct Patches {
+
+  /// \brief Main constructor
+  /// \param patchSize patch size: 2x2, or 4x4
+  /// \param whichSide 0 = A side, 1 = C side
+  /// \param whichSuperModuleSize 0 = Full 1 = 1/3
+  Patches(int patchSize, int whichSide, int whichSuperModuleSize);
+
+  /// \brief Default constructor
+  Patches();
+
+  /// \brief Destructor
+  ~Patches() = default;
+
+  /// \brief Clear internal members
+  void clear();
+
+  /// \brief Initialise internal members
+  void init();
+
+  /// \brief Assign seed module to a Full SM
+  /// \param patchID Patch ID that need to be assigned a seed
+  void assignSeedModuleToPatchWithSTUIndexingFullModule(int& patchID);
+
+  /// \brief Assign seed module to a 1/3 SM
+  /// \param patchID Patch ID that need to be assigned a seed
+  void assignSeedModuleToPatchWithSTUIndexingOneThirdModule(int& patchID);
+
+  /// \brief Assign seed module to all patches
+  void assignSeedModuleToAllPatches();
+
+  /// \brief Assign modules to all patches
+  void assignModulesToAllPatches();
+
+  /// \brief Updates the patches
+  void updateADC();
+
+  int mPatchSize;            //!<! patch size (2x2 or 4x4 typically)
+  int mWhichSide;            //!<! Either A = 0 or C = 1 side
+  int mWhichSuperModuleSize; //!<! Either Full/2/3 = 0 or 1/3 = 1 size
+
+  std::vector<std::tuple<int, int>> mPatchIDSeedFastOrIDs;                  //!<! mask containing Patch IDs, their seed FastOrs
+  std::vector<std::tuple<int, std::vector<int>>> mIndexMapPatch;            //!<! mask of the  FastOrs assigned to each patch
+  std::vector<std::tuple<int, std::vector<int>>> mFiredFastOrIndexMapPatch; //!<! mask of the FastOrs above threshold in each patch
+  std::vector<int> mFiredPatches;                                           //!<! mask of the patches above threshold
+  std::vector<std::tuple<int, std::vector<double>>> mADCvalues;             //!<! ADC values for peak finding
+  std::vector<std::tuple<int, std::vector<double>>> mTimesum;               //!<! Time sums for peak finding
+  std::vector<std::tuple<int, double>> mPreviousTimebinADCvalue;            //!<! ADC that was just removed from the time bins
```

Comment:
Theses are all std::maps. If you can access via a known index (patch index, FastOR index), you better use an array (std::array if size is compile-const or std::vector if the size is assigned dynamically) and access via the index (better, even if some array fields are unused).

Reply:
Unfortunately the way I am doing it is that quite a few of them don't conserve the length of the array during the peak finding. Elements at the beginning of the vectors are erased and pushed back, to avoid hard copies. I would prefer to leave them as they are since I am still using std objects

Reply:
Concerning `mADCvalues`, `mTimesum`,  `mPreviousTimebinADCvalue` trying to switch over to something like `std::array<std::tuple<int, std::deque<double>>>`. The first three/four they are actually fixed size in disguise, trying to switch to `std::array` directly.